pkg/controller/cronhorizontalpodautoscaler/cronjob.go

```diff
@@ -86,60 +90,156 @@ func (ch *CronJobHPA) Ref() *TargetRef {
 }
 
 func (ch *CronJobHPA) Run() (msg string, err error) {
-	targetGK := schema.GroupKind{
-		Group: ch.TargetRef.RefGroup,
-		Kind:  ch.TargetRef.RefKind,
-	}
-	mappings, err := ch.mapper.RESTMappings(targetGK)
-	if err != nil {
-		return "", fmt.Errorf("Failed to create create mapping,because of %s", err.Error())
-	}
 
 	if skip, msg := IsTodayOff(ch.excludeDates); skip {
 		return msg, nil
 	}
 
-	var scale *autoscalingapi.Scale
-	var targetGR schema.GroupResource
-
 	startTime := time.Now()
 	times := 0
 	for {
 		now := time.Now()
 
 		// timeout and exit
 		if startTime.Add(maxRetryTimeout).Before(now) {
-			return "", fmt.Errorf("Failed to scale (namespace: %s;kind: %s;name: %s) to %d after retrying %d times and exit,because of %s", ch.TargetRef.RefNamespace, ch.TargetRef.RefKind, ch.TargetRef.RefName, ch.DesiredSize, times, err.Error())
+			return "", fmt.Errorf("failed to scale (namespace: %s;kind: %s;name: %s) to %d after retrying %d times and exit,because of %s", ch.TargetRef.RefNamespace, ch.TargetRef.RefKind, ch.TargetRef.RefName, ch.DesiredSize, times, err.Error())
 		}
 
-		found := false
-		for _, mapping := range mappings {
-			targetGR = mapping.Resource.GroupResource()
-			scale, err = ch.scaler.Scales(ch.TargetRef.RefNamespace).Get(targetGR, ch.TargetRef.RefName)
-			if err == nil {
-				found = true
+		// hpa compatible
+		if ch.TargetRef.RefKind == "HorizontalPodAutoscaler" {
+			msg, err = ch.ScaleHPA()
+			if err != nil {
+				break
+			}
+		} else {
+			msg, err = ch.ScalePlainRef()
+			if err != nil {
 				break
 			}
 		}
+		time.Sleep(updateRetryInterval)
+		times = times + 1
+	}
+
+	return msg, err
+}
+
+func (ch *CronJobHPA) ScaleHPA() (msg string, err error) {
+	var scale *autoscalingapi.Scale
+	var targetGR schema.GroupResource
+
+	ctx := context.Background()
+	hpa := &autoscalingapi.HorizontalPodAutoscaler{}
+	err = ch.client.Get(ctx, types.NamespacedName{Namespace: ch.HPARef.Namespace, Name: ch.TargetRef.RefName}, hpa)
+
+	if err != nil {
+		return "", fmt.Errorf("Failed to get HorizontalPodAutoscaler Ref,because of %v", err)
+	}
+
+	targetRef := hpa.Spec.ScaleTargetRef
+
+	targetGV, err := schema.ParseGroupVersion(targetRef.APIVersion)
+	if err != nil {
+		return "", fmt.Errorf("Failed to get TargetGroup of HPA %s,because of %v", hpa.Name, err)
+	}
+
+	targetGK := schema.GroupKind{
+		Kind:  targetRef.Kind,
+		Group: targetGV.Group,
+	}
+
+	mappings, err := ch.mapper.RESTMappings(targetGK)
+	if err != nil {
+		return "", fmt.Errorf("Failed to create mapping,because of %s", err.Error())
+	}
+
+	found := false
+	for _, mapping := range mappings {
+		targetGR = mapping.Resource.GroupResource()
+		scale, err = ch.scaler.Scales(ch.TargetRef.RefNamespace).Get(targetGR, ch.TargetRef.RefName)
```

Comment:
`scale, err = ch.scaler.Scales(ch.TargetRef.RefNamespace).Get(targetGR, ch.TargetRef.RefName)` 
 should be
`scale, err = ch.scaler.Scales(ch.TargetRef.RefNamespace).Get(targetGR, hpa.Spec.ScaleTargetRef.Name)` ?